src/librustc_lint/builtin.rs

```diff
@@ -1178,18 +1178,38 @@ fn check_const(cx: &LateContext<'_, '_>, body_id: hir::BodyId) {
     let _ = cx.tcx.const_eval_poly(def_id);
 }
 
+fn check_assoc_const(cx: &LateContext<'_, '_>, body_id: hir::BodyId) {
+    let tcx = cx.tcx;
+    let def_id = tcx.hir().body_owner_def_id(body_id);
+    // Force optimized mir to be loaded, so we get lints even if the associated const is unused.
+    let _mir = tcx.optimized_mir(def_id);
```

Comment:
Huh... This should happen for all associated constants on `pub trait`s even without this PR. I'm not sure I understand how this isn't happening. https://github.com/rust-lang/rust/blob/9bba047c2e425fce03b039bcb8ccd60ddcbc80a0/src/librustc_metadata/rmeta/encoder.rs#L897 should also call this for all exported associated constants.

If it weren't already happening https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d3d29cd2ef2477c0372fb297b785de2d should fail to compile I think.

Maybe the metadata encoding is happening so late that error reporting is already not happening anymore? I remember we used to have some such problems, but I thought those were gone.

Reply:
My theories for why associated consts were not properly linted before involve these two lines, but I am not actually sure what they are doing or why:

https://github.com/rust-lang/rust/blob/d9a328a0ade570608717b0df654306a5268c9f46/src/librustc_mir/transform/const_prop.rs#L602

https://github.com/rust-lang/rust/blob/d9a328a0ade570608717b0df654306a5268c9f46/src/librustc_mir/transform/const_prop.rs#L850

Reply:
> Huh... This should happen for all associated constants on pub traits even without this PR. I'm not sure I understand how this isn't happening.

That entire pass doesn't execute for executables -- the `encode_metadata` query doesn't run for them:

https://github.com/rust-lang/rust/blob/99009bf31992fb9257b84550dbf7cb509f987952/src/librustc_interface/passes.rs#L912-L930

And true enough, if I checkout to before this change and run my test case with `--crate-type=lib` or `--crate-type=dylib`, I get the lint, even with the associated const unused.

Reply:
By "this change", you mean https://github.com/rust-lang/rust/commit/faf5eac854f9099b54854145f2307cae3f3aa768 ?

Reply:
Sorry, I meant before the changes in this PR. Exact commit I went to was a7deedd2db9b65e293a4b7af5bd668f9efba5bfc, so I could run it on my used-unused test case

Reply:
oh... that's actually fairly reasonable then... Dead code isn't even properly type checked in some cases...

We should probably add tests that ensure we emit these lints in `--lib` mode

Reply:
> oh... that's actually fairly reasonable then

In that case, should we preserve the current behaviour of not linting unused associated consts?

> We should probably add tests that ensure we emit these lints in --lib mode

Sure thing. I'll add a new revision to the relevant tests.

Reply:
@oli-obk even as a library with a public crate, [we currently have an issue](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=11c45c0bd4b22ff997e69901b3bb0709).

Reply:
Then I don't understand how

> And true enough, if I checkout to before this change and run my test case with --crate-type=lib or --crate-type=dylib, I get the lint, even with the associated const unused.

can be true. I'm very confused now.

Reply:
@jumbatm on which exact test file did you do the test quoted above? It seems we cannot reproduce that on the playground.

Reply:
`lint-used-unused.rs`, introduced in this PR.(I'd link it but I'm on mobile right now)

Reply:
Ah, that just tests arithmetic overflow. Bitshifts are special because we can lint them without knowing the LHS.
*That* part still does not work in associated consts.

Reply:
We [already have a test for other arithmetic operations](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/issue-69020.rs).

What I don't understand is where that lint is coming from, given that we cannot actually evaluate those consts -- and why it then does not also work for bitshifts.
I think these errors must be coming from const_eval, not const_prop. But that only explains the "why not bitshifts" part, not the "why does it happen at all" part.

Reply:
Ok, so we're back at my question how, for libraries `optimized_mir` is not called for all (even associated) constants (or whether we just don't see the lints during metadata encoding)

Reply:
Hm, looking at [the output](https://github.com/rust-lang/rust/blob/99009bf31992fb9257b84550dbf7cb509f987952/src/test/ui/consts/issue-69020.noopt.stderr) though this the `arithmetic_overflow` lint, not a const_eval panic.

So, it's definitely const_prop. But const_prop bails out early in the case of the bitshift test.
Which brings me back to the two "bail out" lines that I keep mentioning as prime suspects:

https://github.com/rust-lang/rust/blob/d9a328a0ade570608717b0df654306a5268c9f46/src/librustc_mir/transform/const_prop.rs#L601-L603

https://github.com/rust-lang/rust/blob/d9a328a0ade570608717b0df654306a5268c9f46/src/librustc_mir/transform/const_prop.rs#L850

The bug *must* be something where const_prop runs (because we see it run on other, very similar associated consts) but fails to reach the problematic bitshift.

Reply:
some of these bailouts are just papering over interpreter issues, these are likely resolved by now. Or they are there for speedups, which we should benchmark before touching them.

But that still doesn't explain why just running `optimized_mir` creates more lints. Const prop runs as part of `optimized_mir` and not out of any other reason.

Reply:
> But that still doesn't explain why just running optimized_mir creates more lints. Const prop runs as part of optimized_mir and not out of any other reason.

Does it, though? @jumbatm did you confirm that specifically the bitshift testcase from https://github.com/rust-lang/rust/issues/69021 passes that way? (I think we got a lot of confusion because of other testcases, but as the bugreport says the only issue we currently know of is with bitshifts.)

Reply:
See https://github.com/rust-lang/rust/pull/70566#discussion_r401718517: calling `optimized_mir` makes no difference at all, actually.